udp/protocol.go

```diff
@@ -0,0 +1,172 @@
+// Copyright 2015 The Chihaya Authors. All rights reserved.
+// Use of this source code is governed by the BSD 2-Clause license,
+// which can be found in the LICENSE file.
+
+package udp
+
+import (
+	"bytes"
+	"encoding/binary"
+	"net"
+
+	"github.com/chihaya/chihaya/stats"
+	"github.com/chihaya/chihaya/tracker/models"
+)
+
+// initialConnectionID is the magic initial connection ID specified by BEP 15.
+var initialConnectionID = []byte{0, 0, 0x04, 0x17, 0x27, 0x10, 0x19, 0x80}
+
+// eventIDs maps IDs to event names.
+var eventIDs = []string{"", "completed", "started", "stopped"}
+
+var (
+	errMalformedPacket = models.ProtocolError("malformed packet")
+	errMalformedIP     = models.ProtocolError("malformed IP address")
+	errMalformedEvent  = models.ProtocolError("malformed event ID")
+	errBadConnectionID = models.ProtocolError("bad connection ID")
+)
+
+// handleTorrentError writes err to w if err is a models.ClientError.
+func handleTorrentError(err error, w *Writer) {
+	if err == nil {
+		return
+	}
+
+	if models.IsPublicError(err) {
+		w.WriteError(err)
+		stats.RecordEvent(stats.ClientError)
+	}
+}
+
+// handlePacket decodes and processes one UDP request, returning the response.
+func (s *Server) handlePacket(packet []byte, addr *net.UDPAddr) (response []byte, actionName string) {
+	if len(packet) < 16 {
+		return // Malformed, no client packets are less than 16 bytes.
+	}
+
+	connID := packet[0:8]
+	action := binary.BigEndian.Uint32(packet[8:12])
+	transactionID := packet[12:16]
+
+	writer := &Writer{
+		buf: new(bytes.Buffer),
+
+		connectionID:  connID,
+		transactionID: transactionID,
+	}
+
+	defer func() {
+		if writer.buf.Len() > 0 {
+			response = writer.buf.Bytes()
+		}
+	}()
+
+	if action != 0 && !s.connIDGen.Matches(connID, addr.IP) {
+		writer.WriteError(errBadConnectionID)
+		return
+	}
+
+	switch action {
+	case 0:
+		actionName = "connect"
+		if !bytes.Equal(connID, initialConnectionID) {
+			return // Malformed packet.
+		}
+
+		writer.writeHeader(0)
+		writer.buf.Write(s.connIDGen.Generate(addr.IP))
+
+	case 1:
+		actionName = "announce"
+		ann, err := s.newAnnounce(packet, addr.IP)
+
+		if err == nil {
+			err = s.tracker.HandleAnnounce(ann, writer)
+		}
+
+		handleTorrentError(err, writer)
+
+	case 2:
+		actionName = "scrape"
+		scrape, err := s.newScrape(packet)
+
+		if err == nil {
+			err = s.tracker.HandleScrape(scrape, writer)
+		}
+
+		handleTorrentError(err, writer)
+	}
+
+	return
+}
+
+// newAnnounce decodes one announce packet, returning a models.Announce.
+func (s *Server) newAnnounce(packet []byte, ip net.IP) (*models.Announce, error) {
+	if len(packet) < 98 {
+		return nil, errMalformedPacket
+	}
+
+	infohash := packet[16:36]
```

Comment:
I don't think this is really necessary – this is the only place where these values are used (other than potentially in tests), and I don't think they will ever change.